File: 02_LearnDSP-MATLAB/Sec02_FrequencyAnalysis/case01_FFT.m
% Fast Fourier Transformation (FFT)
% Radix-2 Decimation in Time

% XiaoCY 2021-02-01

%%
clear;clc

% X(k) = X1(k) + WN^k X2(k)
% WN(k) = exp(-1i*2*pi*k/N)

%% data
M = 10;
N = 2^M;
x = randn(N,1);                                 % origin data

%% FFT
tic

idx = 0:N-1;                                    % zero-based normal index
iidx = bin2dec(fliplr(dec2bin(idx,M)));         % zero-based inversed index

X = x(iidx+1);                                  % inversed data sequence

for m = 1:M
    Nm = 2^m;
    Nr = Nm/2;
    Wm = exp(-1i*2*pi/Nm);
    Wmk = 1;
    for n = 1:Nr
        for k = n:Nm:N
            kp = k + Nr;
            WX = X(kp)*Wmk;
            X(kp) = X(k)-WX;
            X(k) = X(k)+WX;
        end
        Wmk = Wmk*Wm;
    end
end

toc

%%
Y = fft(x);

ReX = real(X);
ImX = imag(X);
ReY = real(Y);
ImY = imag(Y);

figure
subplot(2,2,1)
plot(idx,ReX)
hold on
plot(idx,ReY,'--')
grid on
xlabel('Index')
ylabel('Real part')

subplot(2,2,2)
plot(idx,ImX)
hold on
plot(idx,ImY,'--')
grid on
xlabel('Index')
ylabel('Imaginary part')

subplot(2,2,3)
plot(idx,ReY-ReX)
grid on
xlabel('Index')
ylabel('Real error')

subplot(2,2,4)
plot(idx,ImY-ImX)
grid on
xlabel('Index')
ylabel('Imaginary error')

File: 03_LearnDSP-Python/Sec02_FrequencyAnalysis/case01_FFT.py
'''
Fast Fourier Transformation (FFT)
Radix-2 Decimation in Time

XiaoCY 2021-02-01
'''

#%%
import numpy as np
import matplotlib.pyplot as plt

def inverse_index(idx,M):
    iidx = idx*0
    for k in range(M):
        iidx += (idx%2) * 2**(M-1-k)
        idx = idx//2
    return iidx

#%% data
M = 10
N = 2**M
x = np.random.randn(N,1)

#%% FFT
idx = np.arange(0,N)
iidx = inverse_index(idx,M)
X = np.array(x[iidx],dtype=complex)

for m in range(1,M+1):
    Nm = 2**m
    Nr = Nm//2
    Wm = np.exp(-1j*2*np.pi/Nm)
    Wmk = 1.
    for n in range(int(Nr)):
        for k in range(n,N,Nm):
            kp = k + Nr
            WX = X[kp] * Wmk
            X[kp] = X[k] - WX
            X[k] = X[k] + WX
        Wmk = Wmk*Wm

#%% check
Y = np.fft.fft(x,axis=0)        # Attention to this 'axis' option

plt.figure
plt.subplot(2,2,1)
plt.plot(idx,X.real)
plt.plot(idx,Y.real,'--')
plt.grid('on')
plt.xlabel('Index')
plt.ylabel('Real part')

plt.subplot(2,2,2)
plt.plot(idx,X.imag)
plt.plot(idx,Y.imag,'--')
plt.grid('on')
plt.xlabel('Index')
plt.ylabel('Imaginary part')

plt.subplot(2,2,3)
plt.plot(idx,Y.real-X.real)
plt.grid('on')
plt.xlabel('Index')
plt.ylabel('Real error')

plt.subplot(2,2,4)
plt.plot(idx,Y.imag-X.imag)
plt.grid('on')
plt.xlabel('Index')
plt.ylabel('Imaginary error')